onnxruntime/contrib_ops/cpu/quantization/matmul_nbits.cc

```diff
@@ -23,27 +24,137 @@
         nbits_{narrow<size_t>(info.GetAttr<int64_t>("bits"))} {
     ORT_ENFORCE(nbits_ == 4,
                 "Only 4b quantization is supported for MatMulNBits op, additional bits support is planned.");
+    info.GetAttrOrDefault<int64_t>("accuracy_level", &accuracy_level_, 0);
+    is_asym_ = info.GetInputCount() >= 4;
   }
 
   Status Compute(OpKernelContext* context) const override;
 
+  Status PrePack(const Tensor& tensor, int input_idx, AllocatorPtr alloc,
+                 /*out*/ bool& is_packed,
+                 /*out*/ PrePackedWeights* prepacked_weights) override;
+
+  Status UseSharedPrePackedBuffers(std::vector<BufferUniquePtr>& prepacked_buffers, int input_idx,
+                                   /*out*/ bool& used_shared_buffers) override;
+
  private:
   const size_t K_;
   const size_t N_;
   const size_t block_size_;
   const size_t nbits_;
   const bool column_wise_quant_{true};
+  IAllocatorUniquePtr<void> packed_b_;
+  size_t packed_b_size_;
+  bool is_asym_;
+  int64_t accuracy_level_;
 };
 
+Status MatMulNBits::PrePack(const Tensor& tensor, int input_idx, /*out*/ AllocatorPtr alloc,
+                            /*out*/ bool& is_packed,
+                            /*out*/ PrePackedWeights* prepacked_weights) {
+  is_packed = false;
+  auto compt_type = static_cast<MLAS_COMPUTE_TYPE>(accuracy_level_);
+  if (MlasNBitsGemmPackBSupport(N_, K_, block_size_, static_cast<int>(nbits_), is_asym_, compt_type)) {
+    // TODO use threadpool here
+    MLAS_THREADPOOL* pool = NULL;
+    if (input_idx == 1) {
+      auto qptr = tensor.Data<uint8_t>();
+      packed_b_size_ = MlasNBitsGemmPackBSize(N_, K_, block_size_, static_cast<int>(nbits_), is_asym_, compt_type);
+      packed_b_ = IAllocator::MakeUniquePtr<void>(alloc, packed_b_size_, true);
+      MlasNBitsGemmPackB(packed_b_.get(), qptr, nullptr, nullptr, N_, K_, K_, block_size_, static_cast<int>(nbits_), is_asym_, false, compt_type, pool);
+      if (prepacked_weights) {
+        prepacked_weights->buffers_.push_back(std::move(packed_b_));
+        prepacked_weights->buffer_sizes_.push_back(packed_b_size_);
+      }
+      is_packed = true;
+    }
+    if (input_idx == 2) {
+      auto sptr = tensor.Data<float>();
+      MlasNBitsGemmPackB(packed_b_.get(), nullptr, sptr, nullptr, N_, K_, K_, block_size_, static_cast<int>(nbits_), is_asym_, !is_asym_, compt_type, pool);
```

Comment:
>packed_b_ [](http://example.com/codeflow?start=25&length=9)

please check packed_b_. If it is nullptr, don't pack for scale and zeropoint

Reply:
fixed in 4c07742421372e23292c0cafd514d8df580fd608